### PR TITLE
remove fill blending for bilinear affine

### LINKED ISSUE #8083: `affine` creates artefacts on the edges of the image
### 🐛 Describe the bug

When employing the affine functional operation (in both v1 and v2), it's evident that black borders are introduced around the image, even when the fill value matches the image content. These black margins are observable when using both uint8 and float32 data types, and this phenomenon occurs consistently on both Ubuntu and Mac M1.

Upon comparing the implementation of the 'affine' operation in torchvision with that in Kornia, I am uncertain whether the interpolation issue is limited to the image edges. Notably, when utilizing Kornia, the output appears to be more visually appealing when applied to an image.

```python
import torch
import torchvision
from torchvision.transforms.v2.functional import affine
from torchvision.tv_tensors import Image
from torchvision.transforms.v2.functional._geometry import _get_inverse_affine_matrix
from kornia.geometry.transform import get_affine_matrix2d, warp_affine
from torchvision.transforms import InterpolationMode

image = Image(128 * torch.ones((3, 240, 200), dtype=torch.float))

angle =30
trans = (0,0)
scale = 1.0
shear = (0,0)
center = (image.shape[-1] / 2, image.shape[-2] / 2)
inter = InterpolationMode.BILINEAR
fill = [128, 128, 128]

M = get_affine_matrix2d(
    torch.Tensor(trans),
    torch.Tensor([center]),
    torch.Tensor([[scale, scale]]),
    torch.Tensor([angle]),
    torch.Tensor([shear[0]]),
    torch.Tensor([shear[1]]),
)

kn_img = warp_affine(
    image.unsqueeze(0),
    M[:, :2],
    image.shape[-2:],
    mode="bilinear",
    padding_mode="fill",
    fill_value=torch.tensor(fill),
    align_corners=False,
)

tv_img = affine(
    image,
    angle=angle,
    translate=trans,
    scale=scale,
    shear=shear,
    fill=fill,
    interpolation=inter,
)

torchvision.io.write_png(kn_img[0].to(dtype=torch.uint8), "affine_kornia.png")
torchvision.io.write_png(tv_img.to(dtype=torch.uint8), "affine_torchvision.png")
```

Kornia | Torchvision
:--:|:--:
affine_kornia |  affine_torchvision

### Versions

PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.6 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.9 (main, Jun 29 2023, 12:23:23) [Clang 14.0.3 (clang-1403.0.22.14.1)] (64-bit runtime)
Python platform: macOS-13.6-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] mypy==1.6.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] onnx==1.14.1
[pip3] pytorch-lightning==2.0.9
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.1.0
[pip3] torch-optimizer==0.3.0
[pip3] torchdata==0.6.1
[pip3] torchmetrics==1.0.3
[pip3] torchtext==0.15.2
[pip3] torchvision==0.16.0

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #8083 and addresses the corresponding part in #6517.

### CODE REVIEW

File: torchvision/transforms/v2/functional/_geometry.py [hunk 1]
```diff
@@ -585,13 +585,8 @@ def _apply_grid_transform(img: torch.Tensor, grid: torch.Tensor, mode: str, fill
...
         fill_img = torch.tensor(fill_list, dtype=float_img.dtype, device=float_img.device).view(1, -1, 1, 1)
-        if mode == "nearest":
-            bool_mask = mask < 0.5
-            float_img[bool_mask] = fill_img.expand_as(float_img)[bool_mask]
-        else:  # 'bilinear'
-            # The following is mathematically equivalent to:
-            # img * mask + (1.0 - mask) * fill = img * mask - fill * mask + fill = mask * (img - fill) + fill
-            float_img = float_img.sub_(fill_img).mul_(mask).add_(fill_img)
+        bool_mask = mask < 1
+        float_img[bool_mask] = fill_img.expand_as(float_img)[bool_mask]
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_geometry.py:589]: This is only equivalent to the 'bilinear' block if `mask` was a boolean `mask`, right?

Is that the case? And if it is, why was there even a distinction between bilinear and nearest mode in the first place?

Comment (pmeier) [reply torchvision/transforms/v2/functional/_geometry.py:589]: > This is only equivalent to the 'bilinear' block if `mask` was a boolean `mask`, right?

Yes.

> Is that the case?

No. 

> And if it is, why was there even a distinction between bilinear and nearest mode in the first place?

I would need to dig, but maybe @vfdev-5 knows? My guess is that the blend strategy we implemented was to have a smooth transition from fill to image. And that works well if we fill with zeros, but creates a shadow artifact if we use a different fill color with bilinear interpolation. See #8083.

Meaning, this PR is BC breaking, but I consider it a bug fix since the blending behavior seems to be "nice to have" while the shadow is an actual issue.

Comment (vfdev-5) [reply torchvision/transforms/v2/functional/_geometry.py:589]: I think it was an incorrect implementation. At the moment of PR (https://github.com/pytorch/vision/pull/2904), I thought it could be ok as assumption to linearly interpolate around the boundaries: https://github.com/pytorch/vision/pull/2904#issuecomment-720548451
